app/lib/shared/deps_graph.dart

```diff
@@ -0,0 +1,218 @@
+    final Map pubspec = pv.pubspec.asJson;
+
+    final Set<String> depsSet = new Set<String>();
+    final normalDeps = pubspec['dependencies'];
+    if (normalDeps is Map) {
+      depsSet.addAll(normalDeps.keys);
+    }
+
+    final Set<String> devDepsSet = new Set<String>();
+    final devDeps = pubspec['dev_dependencies'];
+    if (devDeps is Map) {
+      devDepsSet.addAll(devDeps.keys);
+    }
+
+    // First we add the [package] together with the dependencies /
+    // dev_dependencies to the graph.  This will update the graph transitively,
+    // thereby calculating new users of the [package].
+    _add(package, depsSet, devDepsSet);
+  }
+
+  Set<String> _affectedPackages(String package) {
+    // Due to the constraints in the new [pubspec] it might cause a number of
+    // packages to get new transitive dependencies during a `pub get` and
+    // therefore might cause new analysis results.
+    //
+    // So we trigger all packages that depend (directly or indirectly) on
+    // [package].
+    final Set<String> transitiveUsers =
+        reverseDeps.transitiveNodes(canonicalize(package));
+
+    // We filter out all dev package usages and trigger an analysis for them
+    // (the dev packages are a superset of the normal packages).
+    final Set<String> all = transitiveUsers
+        .where((String p) => p.startsWith(devPrefix))
+        .map((String p) => p.substring(devPrefix.length)).toSet();
+    return all;
+  }
+
+  void _add(String package, Set<String> deps, Set<String> devDeps) {
+    final canonicalizedPackage = canonicalize(package);
+    final canonicalizedDevPackage = canonicalize(_devPackage(package));
+
+    final canonicalizedDeps = deps.map(canonicalize).toSet();
+    final canonicalizedDevDeps = devDeps.map(canonicalize).toSet();
+
+    //normalDeps.addAll(canonicalizedPackage, canonicalizedDeps);
+    //normalDeps.addAll(canonicalizedDevPackage, canonicalizedDeps);
+    for (final String dep in canonicalizedDeps) {
+      reverseDeps.add(dep, canonicalizedPackage);
+      reverseDeps.add(dep, canonicalizedDevPackage);
+    }
+
+    //normalDeps.addAll(canonicalizedDevPackage, canonicalizedDevDeps);
+    for (final String dep in canonicalizedDevDeps) {
+      reverseDeps.add(dep, canonicalizedDevPackage);
+    }
+  }
+
+  // We use a synthetic package to handle weak dependencies / dev
+  // dependencies (no package is depending on this one).
+  String _devPackage(String package) => '$devPrefix$package';
+}
+
+class DepsGraphBasedAnalysisTaskSource implements TaskSource {
+  final DatastoreDB db;
+  final PackageDependencyBuilder builder;
+
+  DepsGraphBasedAnalysisTaskSource(this.db, this.builder);
+
+  @override
+  Stream<Task> startStreaming() async* {
+    await for (final Set<String> packages in builder.run()) {
+      for (final String packageName in packages) {
+        final Key packageKey = db.emptyKey.append(Package, id: packageName);
+        final Package package = (await db.lookup([packageKey]))[0];
```

Comment:
this should be batched by the 10 or something (using e.g. `sliceList` from utils.dart)